### PR TITLE
[MPS] Memory-efficient SDPA backward (opt-in)

### PR DESCRIPTION
**Pull request by christofilojohn**
## Summary

Adds an MPS implementation of `_scaled_dot_product_efficient_attention` and its backward, mirroring the CUDA op signature. The backward never materializes the `[B, H, S, S]` attention matrix that the existing decomposed path requires, it tiles the recomputation along the K/V sequence dimension. The forward saves only `logsumexp` (`[B, H, S]`, fp32), which is enough for the backward to reconstruct softmax row-by-row inside each tile.

The forward is implemented directly with MPSGraph. The backward is implemented in ATen over K/V sequence tiles, the matmul-heavy steps reuse the existing MPS matmul path, which is MPSGraph-backed for floating-point batched matmul. Unsupported cases route back to the dense math path where supported.

## Default behaviour is unchanged

`_fused_sdp_choice_mps` doesn't change the default backend. `SDPBackend.MATH`
is still selected for every shape unless the user explicitly opts in via:

```python
with torch.nn.attention.sdpa_kernel([torch.nn.attention.SDPBackend.EFFICIENT_ATTENTION]):
    out = F.scaled_dot_product_attention(q, k, v)
```

Auto-routing the new backend by default would require benchmark coverage across the M-series SKUs (M1 to M5), which differ in core count, memory bandwidth, and memory size. That work is intentionally out of scope here as I just have an M1 Pro laptop.

## Reasoning

On MPS, training transformers OOMs at long sequence lengths because the backward decomposes to `matmul + softmax_backward + matmul` and materializes the full attention matrix (multiple times in fp32 when inputs are fp16/bf16). CUDA already exposes this efficient-attention operator family and this adds the corresponding MPS implementation path. This PR closes that gap with the same op signature, so the SDP dispatcher can route to it whenever a user opts in.

## Memory results (M1 Pro, B=1, H=4, opt-in path)

| dtype | S    | D   | Math peak | Efficient peak | Reduction |
|-------|------|-----|-----------|----------------|-----------|
| fp16  |  512 | 128 |    50 MB  |     25 MB      | **50%**   |
| fp16  | 1024 |  64 |  1107 MB  |     67 MB      | **94%**   |
| fp16  | 1024 | 128 |  1124 MB  |     50 MB      | **96%**   |
| fp16  | 2048 | 128 |  1116 MB  |    109 MB      | **90%**   |

The reduction grows with `S`, as the algorithm predicts (`O(S²)` → `O(S·BC)`).

Numbers are `torch.mps.driver_allocated_memory()` deltas, i.e. what the MPS allocator reserves end-to-end — the same number that determines whether training OOMs. At S ≥ 1024 the math path's `[B,H,S,S]` fp32 intermediate forces the allocator into its extra-large heap class (~1 GB); the efficient path stays in the smaller heap class. The algorithmic ratio alone is roughly S/BC (≈4–8× for these shapes). The measured allocator reduction can exceed the algorithmic S/BC ratio because avoiding the full fp32 SxS intermediates also avoids larger MPS heap allocations.

## Performance results (M1 Pro, opt-in path vs math)

End-to-end backward (forward + backward + sync), per iter:

| dtype | S    | D   | Math    | Efficient | Speedup vs math                       |
|-------|------|-----|---------|-----------|-----------------------------|
| fp16  |  128 |  64 | 1.14 ms |  1.05 ms  | 1.08× faster |
| fp16  |  128 | 128 | 1.34 ms |  1.05 ms  | 1.29× faster |
| fp16  |  512 |  64 | 1.88 ms |  2.05 ms  | 0.92×   |
| fp16  |  512 | 128 | 1.70 ms |  2.30 ms  | 0.74×   |
| fp16  | 1024 |  64 | 3.35 ms |  4.49 ms  | 0.74× (but 94% memory savings)  |

These numbers are from M1 Pro; other M-series chips have not been measured.

*Default routing is unchanged* so this cannot regress existing workloads, users only see the new path when they ask for it, useful for memory-constrained MPS training workloads (and  the current RAM shortage situation).

## Reproducing the benchmarks

```bash
python benchmarks/transformer/sdpa_mps.py --out results.csv
python benchmarks/transformer/sdpa_mps.py --quick    # short sanity sweep
```

The script depends only on `torch` and writes CSV to stdout (or `--out`).
Contributions of CSVs from other M-series chips are very welcome — that data
is what would unlock auto-routing in a follow-up PR.

## What's in 
- `aten/src/ATen/native/mps/operations/Attention.mm` MPS forward for
  `_scaled_dot_product_efficient_attention`. Returns `(out, logsumexp, philox_seed, philox_offset)`.
- `aten/src/ATen/native/mps/operations/AttentionBackward.mm` — tiled backward via `at::matmul` over K/V slices, plus the dense decomposed fallback for unsupported cases, plus `_fused_sdp_choice_mps` (default = `SDPBackend.MATH`).
- `aten/src/ATen/native/native_functions.yaml` — MPS dispatch entries for `_fused_sdp_choice`, `_scaled_dot_product_efficient_attention`, and `_scaled_dot_product_efficient_attention_backward`.
- `test/test_mps.py` — correctness across dtype × is_causal × shape; dispatcher routing; forward with/without `compute_log_sumexp`
- `benchmarks/transformer/sdpa_mps.py` a single-file CSV benchmark.

## Possible follow-ups

- Default-on routing with a per-SKU heuristic (needs sweeps on chips other than my M1 Pro)
- Fused Metal kernel with `simdgroup_matrix` tiling for additional speedup (would land behind the same dispatch entry this PR adds)
- Dropout support 

## Compatibility

- The default dispatcher still selects `SDPBackend.MATH` for every shape (unchanged).
- The existing decomposed path is unchanged.
- Users who want the memory-efficient path can opt in via `sdpa_kernel([SDPBackend.EFFICIENT_ATTENTION])`, mirroring the CUDA implementation.

Test with:

$ python test/test_mps.py TestSDPA -k efficient

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/Attention.mm [hunk 1]
```diff
@@ -654,6 +655,133 @@ inline bool prefill_attention_supports_head_dim(int64_t head_dim) {
...
+  auto logsumexp = compute_log_sumexp ? at::empty({batchSize, numHeads, qSize}, query.options().dtype(kFloat))
+                                      : at::empty({0}, query.options().dtype(kFloat));
+  auto philox_seed = at::empty({}, query.options().dtype(kLong));
+  auto philox_offset = at::empty({}, query.options().dtype(kLong));
+  const auto scale_factor = sdp::calculate_scale(query, scale).expect_float();
+
+  @autoreleasepool {
+    auto tensor_key = attn_bias.has_value() ? getTensorsStringKey({query, key, value, *attn_bias})
+                                            : getTensorsStringKey({query, key, value});
+    auto key_name = __func__ + tensor_key + ":causal=" + std::to_string(is_causal) +
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/operations/Attention.mm]: Key_name would be cleaner using fmt::format

File: aten/src/ATen/native/mps/operations/Attention.mm [hunk 2]
```diff
@@ -654,6 +655,133 @@ inline bool prefill_attention_supports_head_dim(int64_t head_dim) {
...
+            graph->biasTensor = mpsGraphRankedPlaceHolder(mpsGraph, *attn_bias);
+            scores = [mpsGraph additionWithPrimaryTensor:scores
+                                         secondaryTensor:castMPSTensor(mpsGraph, graph->biasTensor, scores.dataType)
+                                                    name:nil];
+          }
+
+          auto sm = [mpsGraph softMaxWithTensor:scores axis:3 name:nil];
+          auto output = [mpsGraph matrixMultiplicationWithPrimaryTensor:sm secondaryTensor:vTensor name:nil];
+
+          graph->qTensor = qTensor;
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/operations/Attention.mm:748]: Or these qTensor movable? If so std::move should be called on them.

Comment (christofilojohn) [reply aten/src/ATen/native/mps/operations/Attention.mm:748]: These are `MPSGraphTensor*` handles, so `std::move` would only move the pointer value and would not change ownership/lifetime. Plain assignment is intentional here.

File: aten/src/ATen/native/mps/operations/Attention.mm (hunk 1 above)
Comment (christofilojohn) [reply aten/src/ATen/native/mps/operations/Attention.mm]: Agreed, `fmt::format` would make this key construction clearer.
